src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,614 @@
+
+  /** Throws an exception with the given message. */
+  private def illegal(msg: String) = throw new IllegalArgumentException(msg)
+
+  /** The sample barcode metrics in the same order as samples in the sample sheet, with an additional sample for unmatched reads. */
+  private val metrics: Seq[SampleBarcodeMetric] = {
+    val sampleMetrics = samples.map { sample =>
+      val barcode: String = sample.sampleBarcode
+          .getOrElse(illegal(s"Sample with id '${sample.sampleId}' did not have a sample barcode"))
+          .concatenatedBarcode
+      val libraryId: String = sample.libraryId.getOrElse(illegal(s"Sample with id '${sample.sampleId}' did not have a sample barcode"))
+      SampleBarcodeMetric(barcodeName=sample.sampleName, libraryName=libraryId, barcode=barcode)
+    }
+    sampleMetrics.toSeq ++ Seq(SampleBarcodeMetric(UnmatchedSampleId, UnmatchedSampleId, noMatchBarcode))
+  }
+
+  /** Increments the metrics for the given sample ordinal, which matched with the given number of mismatches.  The
+    * sample ordinal less than zero indicates the "unmatched" sample. */
+  def increment(sampleIndex: Int, numMismatches: Int): Unit = {
+    require(0 <= sampleIndex && sampleIndex < this.metrics.length, s"Sample ordinal out of range: $sampleIndex")
+    val metric = metrics(sampleIndex)
+    metric.reads += 1
+    metric.pf_reads += 1
+    if (sampleIndex != this.metrics.length - 1) {
+      if (numMismatches == 0) {
+        metric.perfect_matches += 1
+        metric.pf_perfect_matches += 1
+      }
+      else if (numMismatches == 1) {
+        metric.one_mismatch_matches += 1
+        metric.pf_one_mismatch_matches += 1
+      }
+    }
+  }
+
+  /** Finalize the barcode metrics and return them.  This will return a copy of the barcode metrics. */
+  def barcodeMetrics: Iterable[SampleBarcodeMetric] = {
+    SampleBarcodeMetric.finalizeMetrics(this.metrics.map { m => (m.barcode, m.copy()) }.toMap, this.noMatchBarcode)
+    metrics
+  }
+}
+
+
+private[fastq] object FastqDemultiplexer {
+  /** A class to store the sample ordinal and associated demultiplexed [[SAMRecord]]s.
+    * @param sampleIndex the sample index to which the records were assigned (0-based)
+    * @param numMismatches the # of mismatches if it matched a sample with a sample barcode, -1 otherwise (the 'unmatched'
+    *                      sample).
+    * @param records the records, one for each read that have template bases.
+    */
+  case class DemuxRecord(sampleIndex: Int, numMismatches: Int, records: Seq[SAMRecord])
+
+  /** Counts the nucleotide mismatches between two strings of the same length.  Ignores no calls in expectedBases.
+    * Observed base qualities less than the minimum base quality are counted as mismatches if not a no call.
+    */
+  private[fastq] def countMismatches(observedBases: Array[Byte], expectedBases: Array[Byte]): Int = {
```

Comment:
The optimization in `matchSampleBarcode` is not to even check against _any_ sample barcodes if there are too many no calls in the observed bases.